### PR TITLE
Add `getResourceFiles`

### PR DESCRIPTION
**Pull request by TracerDS**
`getResourceFiles` will return names of all files loaded by a resource

### CODE REVIEW

**Comment by Fernando-A-Rocha** [comment]
Very nice. Can you add an argument "resource element" so you can get file list of another resource, which would work if it is running?

**Comment by TracerDS** [comment]
> Very nice. Can you add an argument "resource element" so you can get file list of another resource, which would work if it is running?
> 

Already did ;)

File: Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp [hunk 1]
```diff
@@ -1468,3 +1470,23 @@ bool CLuaResourceDefs::isResourceProtected(CResource* const resource)
...
+    if (!resource)
+        resource = &lua_getownerresource(luaVM);
+
+    const auto resourceFiles = (*resource)->GetResourceFiles();
+
+    std::vector<std::string> files;
+    files.reserve(resourceFiles.size());
+    for (const auto& file : resourceFiles)
+    {
+        if (file->GetType() != CResourceFile::eResourceType::RESOURCE_FILE_TYPE_CLIENT_FILE)
```
Comment (tederis) [inline Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: Are you sure that a server side function should deal with RESOURCE_FILE_TYPE_CLIENT_FILE files?

Comment (tederis) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: Maybe rename the function to `getLoadedClientFiles` for clarity?

Comment (TracerDS) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: > Maybe rename the function to `getLoadedClientFiles` for clarity?

Doesnt matter. `RESOURCE_FILE_TYPE_CLIENT_FILE` just means a valid `<file>` field. It will return every file thats not a map and not a script.

Comment (tederis) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: `<file>` field means exactly a client file

Comment (tederis) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: And please, do not mark a conversation resolved if the actual conversation is in process.

Comment (TracerDS) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: > `<file>` field means exactly a client file

thats the only file than can be. There is no division between server files and client files because server files is just not a concept.

Comment (tederis) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: > > `<file>` field means exactly a client file
> 
> thats the only file than can be. There is no division between server files and client files because server files is just not a concept.

Exactly for this reason the function should be called `getLoadedClientFiles`(or maybe `getResourceClientFileNames`). Simply because this function cannot deal with server-side files.

Comment (TracerDS) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: Renaming it to `getLoadedClientFiles` would indicate that there is a `getLoadedServerFiles` function. That doesnt make sense because server-side files dont exist.

Comment (tederis) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: > Renaming it to `getLoadedClientFiles` would indicate that there is a `getLoadedServerFiles` function. That doesnt make sense because server-side files dont exist.

The existence of `getLoadedClientFiles` function doesn't lead to the existence of function `getLoadedServerFiles`.

Comment (tederis) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: Anyway, this function must be renamed(maybe you will come up with a better name). Because the current name is confusing.

Comment (Fernando-A-Rocha) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: > Anyway, this function must be renamed(maybe you will come up with a better name). Because the current name is confusing.

I disagree. Instead of renaming it, add a string argument that lets you decide if it should return files, map files, script files, or all. Like "all"/"files"/"scripts"/"maps" according to the xml tags in meta.xml https://wiki.multitheftauto.com/wiki/Meta.xml

Comment (tederis) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: > > Anyway, this function must be renamed(maybe you will come up with a better name). Because the current name is confusing.
> 
> I disagree. Instead of renaming it, add a string argument that lets you decide if it should return files, map files, script files, or all. Like "all"/"files"/"scripts"/"maps" according to the xml tags in meta.xml https://wiki.multitheftauto.com/wiki/Meta.xml

This is what I also come up with. But you got ahead of me. Yes, this will work.

Comment (Fernando-A-Rocha) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: This argument can be optional and default to "all" where it returns script, map and regular file paths all together in a table.

Comment (TheNormalnij) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: `getResourceFiles` is explicit. The name of this function should start with 'getResource' because this function related to resource class. And the function can be shared with this name

Comment (Fernando-A-Rocha) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp]: I agree. That is a more suiting name for this function.

**Review by Fernando-A-Rocha — APPROVED** [review]
LGTM

File: Client/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp [hunk 2]
```diff
@@ -29,6 +30,7 @@ void CLuaResourceDefs::LoadFunctions()
         {"getResourceState", GetResourceState},
         {"loadstring", LoadString},
         {"load", Load},
+        {"getResourceFiles", ArgumentParser<GetResourceFiles>},
```
Comment (TheNormalnij) [inline Client/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:33]: Add field 'files' for the class, please

Comment (TracerDS) [reply Client/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:33]: You mean oop?

Comment (TheNormalnij) [reply Client/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:33]: yeah

File: Server/mods/deathmatch/logic/CResourceFile.h [hunk 3]
```diff
@@ -37,12 +37,22 @@ class CResourceFile
...
 
+    enum class eResourceCategory
+    {
+        ALL = 0,
+        SCRIPTS = 1,
+        MAPS = 2,
+        CONFIGS = 3,
+        FILES = 4,
+    };
+
```
Comment (FileEX) [inline Server/mods/deathmatch/logic/CResourceFile.h]: What is the purpose of setting values ​​here when on the client side the exact same enum has no values ​​set?

Comment (TracerDS) [reply Server/mods/deathmatch/logic/CResourceFile.h]: Forgot to remove it

**Review by Fernando-A-Rocha — APPROVED** [review]

File: Client/mods/deathmatch/logic/CDownloadableResource.cpp [hunk 4]
```diff
@@ -14,6 +14,21 @@
...
+        case eResourceType::RESOURCE_FILE_TYPE_MAP:
+            m_resourceCategoryType = eResourceCategory::MAPS;
+        case eResourceType::RESOURCE_FILE_TYPE_CONFIG:
+        case eResourceType::RESOURCE_FILE_TYPE_CLIENT_CONFIG:
+            m_resourceCategoryType = eResourceCategory::CONFIGS;
+        case eResourceType::RESOURCE_FILE_TYPE_CLIENT_FILE:
+        case eResourceType::RESOURCE_FILE_TYPE_HTML:
+            m_resourceCategoryType = eResourceCategory::FILES;
+    }
+
```
Comment (TheNormalnij) [inline Client/mods/deathmatch/logic/CDownloadableResource.cpp:31]: You can filter it in `CLuaResourceDefs::GetResourceFiles`. You don't need `GetResourceCategoryType`

Comment (TracerDS) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:31]: Its better this way

Comment (TheNormalnij) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:31]: Hmm. You can use original types.  I think `client-scripts` and `server-scripts` have some sense for developers. So you shouldn't map `eResourceType` into custom enum for one Lua function

Comment (TracerDS) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:31]: I'd have to do string manipulation in both functions just to get values.
Its better that way

Comment (TheNormalnij) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:31]: What do you mean?

Comment (TracerDS) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:31]: ArgumentParser handles enum convertion for me. I'd have to accept certain string values and filter them. Its too much hassle.
Its better to create an enum. Everything will be taken care of for me.

Comment (TheNormalnij) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:31]: What is wrong with `eResourceType` ?

Comment (TracerDS) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:31]: too much noise

Comment (TheNormalnij) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:31]: It looks like using existing enum is much better than creating a new one with mapping from `eResourceType` and adding  boilerplate code in client and server.

> too much noise

What do you mean?

Comment (TracerDS) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:31]: > What do you mean?

Exactly what I said. Its easier to have 3 categories: scripts, maps, files instead of: client/server scripts, client/server files, (client/server?) maps, etc.

Comment (TheNormalnij) [reply Client/mods/deathmatch/logic/CDownloadableResource.cpp:31]: I see no problem here. It's much useful.

**Comment by TheNormalnij** [comment]
Does it return shared scripts twice here? Is it expected behavior?
```lua
local list = getResourceFiles("all")
```

**Comment by Fernando-A-Rocha** [comment]
File paths returned shouldn't be repeated imo

**Comment by Fernando-A-Rocha** [comment]
To make this more advanced, the list returned could be a table of {filePath, fileType} where fileType is like "client-script" or "shared-script" or "map-file" idk

**Comment by TracerDS** [comment]
> Does it return shared scripts twice here? Is it expected behavior?
> 
> ```lua
> local list = getResourceFiles("all")
> ```

Yeah, it returns twice. Its not expected behaviour. Will fix that
image

File: Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp [hunk 5]
```diff
@@ -1476,24 +1476,28 @@ bool CLuaResourceDefs::isResourceProtected(CResource* const resource)
...
     for (const auto& file : resourceFiles)
     {
-        if (file->GetResourceCategoryType() == type.value()
-            || type.value() == CResourceFile::eResourceCategory::ALL)
-        {
-            files.push_back(file->GetName());
-        }
+        if (file->GetResourceCategoryType() != type.value() && type.value() != eResourceCategory::ALL)
+            continue;
+        files.insert(file->GetName());
```
Comment (TheNormalnij) [inline Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: Previous varian was more readable. Why did you change it?

Comment (TracerDS) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: previous variant was basically the same.
This leaves room for more conditions to come.

Comment (TheNormalnij) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: There is no 'early continue' rule. You obfuscate the code with overusing 'continue' in loops. And you don't really need to use early return for small code blocks.
You did the code less readable and maintainable.
```c++
if (file->GetResourceCategoryType() == type.value() || type.value() == eResourceCategory::ALL)
    files.insert(file->GetName());
```
Is fine condition, that describes the function logic.

Comment (TracerDS) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: There is early-return which is basically the same but whatever.
I'd rather go with continue rather than indentation

Comment (TheNormalnij) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: Your `continue` creates indentation too. Please, keep it simple

> There is early-return which is basically the same but whatever.

Not really. Multiple 'continue' conditions can make a loop harder for reading.
There is no goal to use minimal indentation. The code should primarily describe the logic.
You're overusing 'early return'

Comment (TheNormalnij) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: Don't resolve conversations without changes silently

Comment (TracerDS) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: Im not going to change that. We didnt continue the discussion for some time, hence I resolved it.
Conditions may change. Better to overuse early return than to not use it at all/not use it often enough.
Result? Mta code with huge if blocks.

Comment (TheNormalnij) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: > We didnt continue the discussion for some time, hence I resolved it.

It looks like you tried to hide requested changes. Be civil. Maybe another mta member will points that I'm wrong or confirms my position.

> Conditions may change.

This is not how overused code looks like. It's better to keep it readable for future changes.

> Better to overuse early return than to not use it at all/not use it often enough.

You deal in absolutes. Overused principles are harmful for readability and complexity. Use them only when they improve your code. In this example you trade readability for principles.

Comment (TracerDS) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: > > We didnt continue the discussion for some time, hence I resolved it.
> 
> It looks like you tried to hide requested changes. Be civil. Maybe another mta member will points that I'm wrong or confirms my position.
> 
> > Conditions may change.
> 
> This is not how overused code looks like. It's better to keep it readable for future changes.
> 
> > Better to overuse early return than to not use it at all/not use it often enough.
> 
> You deal in absolutes. Overused principles are harmful for readability and complexity. Use them only when they improve your code. In this example you trade readability for principles.

If more contributors are opposed against this change, I will reconsider

**Review by Fernando-A-Rocha — APPROVED** [review]
Ready to merge imo

File: Server/mods/deathmatch/logic/CResource.h [hunk 6]
```diff
@@ -256,8 +256,8 @@ class CResource : public EHS
     const std::string&          GetResourceDirectoryPath() const { return m_strResourceDirectoryPath; }
     const std::string&          GetResourceCacheDirectoryPath() const { return m_strResourceCachePath; }
 
-    std::list<CResourceFile*>& GetFiles() { return m_ResourceFiles; }
-    size_t                     GetFileCount() const noexcept { return m_ResourceFiles.size(); }
+    std::list<CResourceFile*>&       GetResourceFiles() noexcept { return m_ResourceFiles; }
```
Comment (tederis) [inline Server/mods/deathmatch/logic/CResource.h:259]: Non-const getter is usually a bad practice. Is it really needed here?

File: Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp (hunk 5 above)
Comment (TheNormalnij) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: Again?

File: Server/mods/deathmatch/logic/CResource.h (hunk 6 above)
Comment (TracerDS) [reply Server/mods/deathmatch/logic/CResource.h:259]: I thought it may be modified somewhere. Better safe than sorry.
If it wont be modified anywhere then I can remove it

File: Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp (hunk 5 above)
Comment (TracerDS) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: > Again?

No one was interested in this convo for 2 weeks, give it a break.
No contributors were opposed against this so Im leaving it as it is.

Comment (tederis) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: I also see the early return overuse here.

File: Server/mods/deathmatch/logic/CResource.h (hunk 6 above)
Comment (tederis) [reply Server/mods/deathmatch/logic/CResource.h:259]: I don't think that this getter will be useful. Even if so, it can be recreated very easy.

File: Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp (hunk 5 above)
Comment (TheNormalnij) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: > No contributors were opposed against this so Im leaving it as it is.

And nobody confirm your position. Don't hide requested changes. Stale conversations should be dismissed by another  members.

Comment (TracerDS) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: > And nobody confirm your position

Im following the guidelines is'all.

Comment (TheNormalnij) [reply Server/mods/deathmatch/logic/luadefs/CLuaResourceDefs.cpp:1494]: you interpreter these guidelines very differently

**Comment by Fernando-A-Rocha** [comment]
@TracerDS it would be nice to have a way to know if a file has download="false" attribute, so it can be downloaded using downloadFile later
Currently the only way to achieve this is by parsing the meta.xml file, which seems a bit too much

**Comment by Fernando-A-Rocha** [comment]
> @TracerDS it would be nice to have a way to know if a file has download="false" attribute, so it can be downloaded using downloadFile later Currently the only way to achieve this is by parsing the meta.xml file, which seems a bit too much

Any ideas?

**Comment by TracerDS** [comment]
> > @TracerDS it would be nice to have a way to know if a file has download="false" attribute, so it can be downloaded using downloadFile later Currently the only way to achieve this is by parsing the meta.xml file, which seems a bit too much
> 
> Any ideas?

maybe a parameter that checks download="false"?

**Comment by Fernando-A-Rocha** [comment]
> > > @TracerDS it would be nice to have a way to know if a file has download="false" attribute, so it can be downloaded using downloadFile later Currently the only way to achieve this is by parsing the meta.xml file, which seems a bit too much
> > 
> > 
> > Any ideas?
> 
> maybe a parameter that checks download="false"?

Resource files, when they are parsed in the meta.xml, are already identified if they need to be downloaded later by the client.

**Comment by Fernando-A-Rocha** [comment]
It's a matter of returning that download attribute value for each file. The array of file names returned by this new function should contain arrays where you have like
`{fileName="myFoder/myFile.dff", download=false}`

**Comment by TracerDS** [comment]
> It's a matter of returning that download attribute value for each file. The array of file names returned by this new function should contain arrays where you have like
> `{fileName="myFoder/myFile.dff", download=false}`

I think it would be better to return just an array of files and handle the download attribute via a parameter. It would be better for the end user

**Comment by Fernando-A-Rocha** [comment]
> > It's a matter of returning that download attribute value for each file. The array of file names returned by this new function should contain arrays where you have like
> > `{fileName="myFoder/myFile.dff", download=false}`
> 
> I think it would be better to return just an array of files and handle the download attribute via a parameter. It would be better for the end user

True, but what parameter exactly?

**Comment by TracerDS** [comment]
> True, but what parameter exactly?

Something like isDownloaded or something like that. Or maybe it will receive a table of flags if more param were to come

**Comment by TracerDS** [comment]
Im not interested in advancing this PR further anymore.
If someone wants to refresh it, feel free to do so.
